src/Tools/dotnet-monitor/Exceptions/ExceptionsStore.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using Microsoft.Diagnostics.Monitoring.WebApi.Exceptions;
+using Microsoft.Diagnostics.Monitoring.WebApi.Stacks;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Exceptions
+{
+    internal sealed class ExceptionsStore : IExceptionsStore, IAsyncDisposable
+    {
+        private const int ChannelCapacity = 1000;
+
+        private readonly Channel<ExceptionInstanceEntry> _channel;
+        private readonly CancellationTokenSource _disposalSource = new();
+        private readonly List<ExceptionInstance> _instances = new();
+        private readonly Task _processingTask;
+
+        private long _disposalState;
+
+        public ExceptionsStore()
+        {
+            _channel = CreateChannel();
+            _processingTask = ProcessEntriesAsync(_disposalSource.Token);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (!DisposableHelper.CanDispose(ref _disposalState))
+                return;
+
+            _channel.Writer.TryComplete();
+
+            await _processingTask.SafeAwait();
+
+            _disposalSource.SafeCancel();
+
+            _disposalSource.Dispose();
+        }
+
+        public void AddExceptionInstance(IExceptionsNameCache cache, ulong exceptionId, string message)
+        {
+            ExceptionInstanceEntry entry = new(cache, exceptionId, message);
+            // This should never fail to write because the behavior is to drop the oldest.
+            _channel.Writer.TryWrite(entry);
+        }
+
+        public IEnumerable<IExceptionInstance> GetSnapshot()
```

Comment:
nit: Consider returning IList instead of IEnumerable. IEnumerable signals deferred execution at the cost of random access. We really can't do deferred computation here so we might as well provide random access abilities.